File: simpcitybot.py
import time
import random
import os
from instagrapi import Client
from dotenv import load_dotenv

# Load secret variables from .env file
load_dotenv()

USERNAME = os.getenv("INSTA_USERNAME")
PASSWORD = os.getenv("INSTA_PASSWORD")
GROUP_THREAD_ID = os.getenv("GROUP_THREAD_ID")

# Welcome Messages List
WELCOME_MESSAGES = [
    "Welcome @{username} to SimpCity! 🚀🔥",
    "Aree bhai! Ek aur simp aaya! 😆 Welcome @{username}! 🎉",
    "@{username} just entered the simp zone. Buckle up! 😂",
    "Ek naye simp ki entry hui hai! @{username}, welcome bhai! 💀🔥"
]

# Login to Instagram
cl = Client()
cl.login(USERNAME, PASSWORD)

def get_group_members():
    """Fetch group members and return as dictionary"""
    try:
        group_info = cl.direct_thread(GROUP_THREAD_ID)
        members = {user.pk: user.username for user in group_info.users}
        print("🟢 Current Group Members:", members)  # Debugging line
        return members
    except Exception as e:
        print(f"❌ Error fetching group members: {e}")
        return {}
    
def mention_all():
    """Mentions all group members"""
    members = get_group_members()
    if not members:
        return "⚠ No members found!"
    
    mentions = " ".join([f"@{username}" for username in members.values()])
    message = f"🚀 Mentioning everyone: {mentions}"
    
    cl.direct_send(text=message, thread_ids=[GROUP_THREAD_ID])
    print(f"✅ Mentioned all members: {mentions}")

def check_messages():
    """Checks latest group messages & triggers mention if needed"""
    try:
        messages = cl.direct_messages(GROUP_THREAD_ID)
        latest_message = messages[0].text.lower() if messages else ""

        if "mention all" in latest_message:
            mention_all()

    except Exception as e:
        print(f"❌ Error checking messages: {e}")

# Track previous members
previous_members = get_group_members()

while True:
    check_messages()
    current_members = get_group_members()
    
    # Detect new members
    new_member_ids = set(current_members.keys()) - set(previous_members.keys())

    for member_id in new_member_ids:
        username = current_members[member_id]  # Fetch username from dictionary
        message = random.choice(WELCOME_MESSAGES).replace("{username}", username)
        cl.direct_send(text=message, thread_ids=[GROUP_THREAD_ID])
        print(f"✅ Welcomed {username} in group chat!")

    previous_members = current_members  # Update members list

    # Check every 10 seconds
    time.sleep(5)

File: simpcitybot.js
require("dotenv").config(); // Load environment variables
const axios = require("axios");

const GROUP_IDS = process.env.GROUP_IDS.split(","); // Multiple group IDs (comma separated)
const ACCESS_TOKEN = process.env.ACCESS_TOKEN;
const INTERVAL = 60 * 1000; // 1 minute

let knownMembers = {}; // Store previous members list

// Function to fetch group members
async function getGroupMembers(groupId) {
  try {
    const url = `https://graph.facebook.com/v18.0/${groupId}/participants?access_token=${ACCESS_TOKEN}`;
    const response = await axios.get(url);
    return response.data.data.map((member) => member.id);
  } catch (error) {
    console.error(
      `Error fetching members for Group ${groupId}:`,
      error.response?.data || error.message
    );
    return [];
  }
}

// Function to welcome new members
async function welcomeNewMembers(groupId, newMembers) {
  for (const memberId of newMembers) {
    const message = `Welcome <@${memberId}> to SimpCity! 🎉👀 Drop a message and introduce yourself!`;
    await sendMessage(groupId, message);
  }
}

// Function to send a message
async function sendMessage(groupId, message) {
  try {
    const url = `https://graph.facebook.com/v18.0/${groupId}/messages?access_token=${ACCESS_TOKEN}`;
    await axios.post(url, { message });
    console.log(`✅ Sent welcome message to Group ${groupId}`);
  } catch (error) {
    console.error(
      `❌ Error sending message:`,
      error.response?.data || error.message
    );
  }
}

// Function to check for new members
async function checkForNewMembers(groupId) {
  const members = await getGroupMembers(groupId);

  if (!knownMembers[groupId]) {
    knownMembers[groupId] = members; // First time setup
    return;
  }

  const newMembers = members.filter(
    (member) => !knownMembers[groupId].includes(member)
  );

  if (newMembers.length > 0) {
    console.log(`🎉 New members detected in Group ${groupId}:`, newMembers);
    await welcomeNewMembers(groupId, newMembers);
  }

  knownMembers[groupId] = members; // Update known members list
}

// Function to check all groups every 1 minute
async function monitorGroups() {
  for (const groupId of GROUP_IDS) {
    await checkForNewMembers(groupId);
  }
}

// Start checking every 1 minute
setInterval(monitorGroups, INTERVAL);

console.log("🚀 Auto-Welcome Bot is Running...");
